fedimint-dbtool/src/main.rs

```diff
@@ -0,0 +1,73 @@
+use anyhow::Result;
+use bitcoin_hashes::hex::ToHex;
+use bytes::Bytes;
+use clap::{Parser, Subcommand};
+use fedimint_api::db::Database;
+use fedimint_api::module::registry::ModuleDecoderRegistry;
+
+#[derive(Debug, Clone, Parser)]
+struct Options {
+    database: String,
+    #[command(subcommand)]
+    command: DbCommand,
+}
+
+#[derive(Debug, Clone, Subcommand)]
+enum DbCommand {
+    List {
+        #[arg(value_parser = hex_parser)]
+        prefix: Bytes,
+    },
+    Write {
+        #[arg(value_parser = hex_parser)]
+        key: Bytes,
+        #[arg(value_parser = hex_parser)]
+        value: Bytes,
+    },
+    Delete {
+        #[arg(value_parser = hex_parser)]
+        prefix: Bytes,
+    },
+}
+
+fn hex_parser(hex: &str) -> Result<Bytes> {
+    let bytes: Vec<u8> = bitcoin_hashes::hex::FromHex::from_hex(hex)?;
+    Ok(bytes.into())
+}
+
+async fn open_db(path: &str) -> Result<Database> {
+    let rocksdb = fedimint_rocksdb::RocksDb::open(path)?;
+    Ok(Database::new(rocksdb, ModuleDecoderRegistry::default()))
+}
+
+fn print_kv(key: &[u8], value: &[u8]) {
+    println!("{} {}", key.to_hex(), value.to_hex());
+}
+
+#[tokio::main]
+async fn main() {
+    let options: Options = Options::parse();
+
+    let db = open_db(&options.database).await.expect("Failed to open DB");
+    let mut dbtx = db.begin_transaction().await;
+
+    match options.command {
+        DbCommand::List { prefix } => {
+            let mut prefix_iter = dbtx.raw_find_by_prefix(&prefix).await;
+            for db_res in prefix_iter.next() {
+                let (key, value) = db_res.expect("DB error");
+                print_kv(&key, &value);
+            }
+        }
+        DbCommand::Write { key, value } => {
+            dbtx.raw_insert_bytes(&key, value.into())
+                .await
+                .expect("DB error");
+        }
+        DbCommand::Delete { prefix: key } => {
+            dbtx.raw_remove_entry(&key).await.expect("DB error");
+        }
+    }
+
+    dbtx.commit_tx().await.expect("DB error");
```

Comment:
Hmm I'm not seeing anything obvious on why this wouldn't be persisting values. If you use this tool with a new, empty db, does that work? Is commit crashing?

Reply:
Lol, the commit tracking is only implemented in the "normal", get/delete/… functions and not the `raw_*` versions.

That shouldn't be a problem, but apparently commits are only done if the commit tracker thinks something happened, which is an unnecessary and imo dangerous optimization. Fixed in 9a6f594c3db59379b8aaecf348a3e2f2317cf21a.